### DISCUSSION TITLE
VScode Plugin for JSONNET

### DISCUSSION THREAD

**Discussion by shibumi** [grafana/tanka#503] [Q&A] [answered]
Hi,
Do you have any recommendation for syntax highlighting etc for jsonnet for vscode?
I searched for it in the plugin store and I couldn't find one.

![https://kurisu.shibumi.dev/share/img-2021-02-04-17-29-46.png](https://kurisu.shibumi.dev/share/img-2021-02-04-17-29-46.png)

**Comment by Duologic** [comment, marked as answer]
Have a look at this: https://marketplace.visualstudio.com/items?itemName=heptio.jsonnet

**Reply by sbarzowski** [reply]
Unfortunately it's pretty much abandoned at this point. It would be really nice if we had an "official" VS Code extension. I'll be happy to help if someone wants to work on this.

**Reply by Duologic** [reply]
I think you can fork it, I don't know much about licensing tho. I'm sure more VScode users would be interested in a maintained version.

**Reply by tombrk** [reply]
It's Apache v2 licensed: https://github.com/heptio/vscode-jsonnet/blob/master/LICENSE, so a fork should be no problem after all.

As this is requested very frequently, this is for sure something one could argue falls into the interest domain of Tanka team.
The existing plugin gives syntax highlighting, but this plugin attempts to do more, e.g. it seems to have a TS implementation of a Jsonnet parser: https://github.com/heptio/vscode-jsonnet/tree/master/compiler

I'd very interested on hearing @sbarzowski thoughts on how Jsonnet code analysis / Language Server features could be done:
- Getdoc (Documentation on hover, maybe using [docsonnet](https://github.com/jsonnet-libs/docsonnet)?
- Godef (Jump to definition). Tricky part appears to figure out what the ultimate source is, after all overriding happened. Is this even possible without running the full evaluation (which takes long)?
- Complete: Suggest possible symbols based on current cursor position

Given single Jsonnet files are not self-contained (they can possibly part of a chain of overrides), those would need to be context aware. For Tanka this is easy, one would need to specify the environment and we can figure out the rest.

**Reply by sbarzowski** [reply]
My approach would be to focus on getting the basics right in the beginning. Getting the highlighting right etc. This extension should just work after installation. The existing extension has fancy features, but the basics are problematic (it reports errors where there are none etc.).

Once we have that, we can gradually add more advanced features. The most maintainable option is to take advantage of the official implementation. I certainly don't want to maintain what is basically another implementation of Jsonnet in TypeScript.

**Reply by davidovich** [reply]
Maybe this is too experimental, but what about implementing the extension in go and use webassembly for the extension ?

https://john-millikin.com/extending-vscode-with-webassembly

This would allow feature parity with the golang version, without having to maintain the TypeScript version.

Just an idea, which needs some work to prove doable.

**Reply by sbarzowski** [reply]
Yeah, we can probably do that for the fancy features. We can also use language server protocol and have it running in a separate process (and have it potentially usable with other editors as well).

For the basic version, I wouldn't be surprised if some simple configuration (grammar description etc.) would suffice, without any custom code.

**Reply by davidovich** [reply]
Yes, maybe the language server protocol is probably a better implementation idea, as we have seen with gopls, many editors have benefited from that type of decoupling.

**Comment by Almenon** [comment]
I've been using https://marketplace.visualstudio.com/items?itemName=Grafana.vscode-jsonnet. So far it's a bit buggy but it works okay.
